oslo/transformers/trainer.py

```diff
@@ -0,0 +1,852 @@
+
+    def _prepare_input(self, data: Union[torch.Tensor, Any]) -> Union[torch.Tensor, Any]:
+        """
+        Prepares one `data` before feeding it to the model, be it a tensor or a nested list/dictionary of tensors.
+        """
+        if isinstance(data, Mapping):
+            return type(data)({k: self._prepare_input(v) for k, v in data.items()})
+        elif isinstance(data, (tuple, list)):
+            return type(data)(self._prepare_input(v) for v in data)
+        elif isinstance(data, torch.Tensor):
+            kwargs = dict(device=self.args.device)
+            return data.to(**kwargs)
+        return data
+
+    def _prepare_inputs(self, inputs: Dict[str, Union[torch.Tensor, Any]]) -> Dict[str, Union[torch.Tensor, Any]]:
+        """
+        Prepare `inputs` before feeding them to the model, converting them to tensors if they are not already and
+        handling potential state.
+        """
+        inputs = self._prepare_input(inputs)
+        if len(inputs) == 0:
+            raise ValueError(
+                "The batch received was empty, your model won't be able to train on it. Double-check that your "
+                f"training dataset contains keys expected by the model."
+            )
+        # TODO mems
+        # if self.args.past_index >= 0 and self._past is not None:
+        #     inputs["mems"] = self._past
+
+        return inputs
+
+    def _get_train_sampler(self) -> Optional[torch.utils.data.Sampler]:
+        generator = None
+        # TODO
+        # if self.args.world_size <= 1 and _is_torch_generator_available:
+        #     generator = torch.Generator()
+        #     # for backwards compatibility, we generate a seed here (which is sampled from a generator seeded with
+        #     # `args.seed`) if data_seed isn't provided.
+        #     # Further on in this method, we default to `args.seed` instead.
+        #     if self.args.data_seed is None:
+        #         seed = int(torch.empty((), dtype=torch.int64).random_().item())
+        #     else:
+        #         seed = self.args.data_seed
+        #     generator.manual_seed(seed)
+        #
+        # seed = self.args.data_seed if self.args.data_seed is not None else self.args.seed
+
+        if self.args.world_size > 1 and self.args.oslo_config.data_parallelism is not None:
+            if not self.args.dataloader_drop_last:
+                return DistributedSamplerWithLoop(
+                    self.train_dataset,
+                    batch_size=self.args.per_device_train_batch_size,
+                    num_replicas=self.args.world_size,
+                    rank=self.args.process_index,
+                    # seed=seed, TODO oslo seed
+                )
+            else:
+                return DistributedSampler(
+                    self.train_dataset,
+                    num_replicas=self.args.world_size,
+                    rank=self.args.process_index,
+                    # seed=seed, TODO oslo seed
```

Comment:
same with above.

Reply:
```python
num_replicas=self.parallel_context.get_world_size(ParallelMode.DATA)
rank=self.parallel_context.get_local_rank(ParallelMode.DATA)
```

Reply:
Instead duplicate self.parallel_context.get_local_rank for each sampler or other place, set them on Training Arguments